lib/ArrayInterfaceCore/src/ArrayInterfaceCore.jl

```diff
@@ -562,6 +557,29 @@ ndims_shape(@nospecialize T::Type{<:Number}) = 0
 ndims_shape(@nospecialize T::Type{<:AbstractArray}) = ndims(T)
 ndims_shape(x) = ndims_shape(typeof(x))
 
+"""
+    IndicesInfo(T::Type{<:Tuple}) -> IndicesInfo{NI,NS,IS}()
+
+Provides basic trait information for each index type in in the tuple `T`. `NI`, `NS`, and
+`IS` are tuples of [`ndims_index`](@ref), [`ndims_shape`](@ref), and
+[`is_splat_index`](@ref) (respectively) for each field of `T`.
+"""
+struct IndicesInfo{NI,NS,IS} end
+IndicesInfo(@nospecialize x::Tuple) = IndicesInfo(typeof(x))
+@generated function IndicesInfo(::Type{T}) where {T<:Tuple}
+    NI = Expr(:tuple)
+    NS = Expr(:tuple)
+    IS = Expr(:tuple)
+    for i in 1:fieldcount(T)
+        T_i = fieldtype(T, i)
+        push!(NI.args, :(ndims_index($(T_i))))
+        push!(NS.args, :(ndims_shape($(T_i))))
+        push!(IS.args, :(is_splat_index($(T_i))))
+    end
+    :(IndicesInfo{$(NI),$(NS),$(IS)}())
+end
```

Comment:
Seems unfortunate that this sort of thing has to be a generated function.
If we have a lot of code that wants to do something like this, it may be worth defining a `typetail` to take a recursive approach.

Reply:
The generated function itself isn't costing very much in load time but if we end up needing to do the same thing a lot we could have a single generated `each_field(f, ::Type{<:Tuple})` to avoid defining too many of these. What we need is essentially `ntuple(i -> f(fieldtype(T, i)), Val{fieldcount(T)}())` but I've noticed constant propagation gives up pretty easily with `fieldtype`.

Either way, I should at least try being proactive about this before merging so I'll try a few things out before merging this.

Reply:
While I do think latency is a major issue, I am more worried about const prop failures.
(And many latency problems do come from type instabilities.)

Reply:
I tried using this code instead:
```julia
struct IndexInfo{NI,NS,IS} end
IndexInfo(x) = IndexInfo{ndims_index(x),ndims_shape(x),is_splat_index(x)}()

indices_info(@nospecialize x::Tuple) = indices_info(typeof(x))
function indices_info(::Type{T}) where {T<:Tuple}
    ntuple(i -> IndexInfo(fieldtype(T, i)), Val{fieldcount(T)}())
end
```

and inference was fine. I'm not completely confident it will hold up in practice because at the very least it will be used for `ArrayInterface._to_indices` which is already several steps away from the call site `ArrayInterface.getindex`.

If there's something else I'm missing I'm open to trying it out.